x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/optimizer/Optimizer.java

```diff
@@ -86,14 +82,10 @@ public LogicalPlan optimize(LogicalPlan verified) {
         Batch operators = new Batch("Operator Optimization",
                 new ConstantFolding(),
                 // boolean
-                new BooleanSimplification(),
+                new EqlBooleanSimplification(),
                 new LiteralsOnTheRight(),
                 new BinaryComparisonSimplification(),
                 new BooleanFunctionEqualsElimination(),
```

Comment:
It's possible that this optimization could lead to inconsistent results in a multi-valued context, so we may want to investigate further.
```
field = ["foo", "bar"]

(field == "foo") == true // should be evaluated as true
(field == "foo") == false // true? or false?
```

I think this optimization is okay in the single-valued context, like in the painless lambda. This is probably similar to the `string(x == y and ... or ...)` situation and the likelihood of a typical person using EQL this way is fairly low, IMHO.

Reply:
`string(x == y and .....)` is, indeed, something that doesn't probably work as expected at the moment when `not` is also used and I'd be happy removing support for `string(Boolean)` if it doesn't have too much sense or real life usage.
`field == "foo"` is always `true`.
`field == "bar"` is always `true`.
(pseudocode) `field != "foo"` -> `not field == "foo"` -> `not (any field == "foo")`/`all (field != "foo")` -> `false`

@rw-access do you have a more concrete example in the form of a real eql query?